src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala

```diff
@@ -258,4 +266,57 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
       postResponse.entity.asString.contains(Conflict.reason)
   }
 
+  // get a mapping of FireCloud user name to NIH User name, for only those Thurloe users with a non-expired NIH link
+  def getCurrentNihUsernameMap: Future[Map[String, String]] = {
+
+    val pipeline = addCredentials(OAuth2BearerToken(HttpGoogleServicesDAO.getAdminUserAccessToken)) ~> sendReceive
+    def getAllUserValuesForKey(key: String): Future[Map[String, String]] = pipeline {
+      Get(s"${UserService.remoteGetQueryURL}key=${key}")
+    } map { response =>
+      if (response.status != OK) throw new Exception(response.toString)   // TODO: better error handling
+      response.entity.as[Seq[ThurloeKeyValue]] match {
+        case Right(seq) =>
+          val resultOptions = seq map { tkv => (tkv.userId, tkv.keyValuePair.flatMap { kvp => kvp.key }) }
+          val actualResultsOnly = resultOptions collect { case (Some(firecloudSubjId), Some(thurloeValue)) => (firecloudSubjId, thurloeValue) }
+          actualResultsOnly.toMap
+        case Left(err) => throw new Exception(err.toString)   // TODO: better error handling
+      }
+    }
+
+    val usernames = getAllUserValuesForKey("linkedNihUsername")
+    val expiretimes = getAllUserValuesForKey("linkExpireTime")
+
+    // compose futures
+    for {
+      usernameKvp <- usernames
+      expirationKvp <- expiretimes
+    } yield {
+      // exclude from results: link expiration is not in Thurloe, value is not a long, value is in the past
```

Comment:
I am not entirely certain that the below works. At the very least all the flatMaps make it hard to follow. What I would probably do is first find all the non-expired fc users:

val validFcUsers = expirationKvp.map{ case (fcUser, expStr) => fcUser -> Try { expStr.toLong }.toOption }.collect {
  case (fcUser, Some(exp: Long)) if DateUtils.now > exp => fcUser
}

Then filter usernameKvp with the results from above:
usernameKvp.filter { case (fcUser, nihUser) => validFcUsers.contains(fcUser) } values

(maybe that all compiles)